pysb/simulator/scipyode.py

```diff
@@ -551,6 +417,353 @@ def _integrator_process(initials, param_values, code_eqs, jac_eqs, num_species,
+            for i, _ in enumerate(model.species)
+        }
+        param_subs = dict(zip(model.parameters_all(), self.p))
+        obs_subs = dict(zip(model.observables, self.o))
+        expr_dynamic_subs = {
+            e: e.expand_expr()
+            for i, e in enumerate(expr_dynamic)
+        }
+        expr_constant_subs = dict(zip(expr_constant, self.e))
+        obs_matrix = scipy.sparse.lil_matrix(
+            (len(model.observables), len(model.species)), dtype=np.int64
+        )
+        for i, obs in enumerate(model.observables):
+            obs_matrix[i, obs.species] = obs.coefficients
+        self.kinetics = sympy.Matrix([
+            r['rate']
+            .xreplace(expr_constant_subs)
+            .xreplace(expr_dynamic_subs)
+            .xreplace(param_subs)
+            .xreplace(species_subs)
+            .xreplace(obs_subs)
+            for r in model.reactions
+        ])
+        if with_jacobian:
+            # The Jacobian can be quite large but it's extremely sparse. We can
+            # obtain a sparse representation by making a SparseMatrix copy of
+            # the kinetics vector and computing the Jacobian on that.
+            kinetics_sparse = sympy.SparseMatrix(self.kinetics)
+            # Rather than substitute all the observables linear-combination-of-
+            # species expressions into the kinetics and then compute the
+            # Jacobian, we can use the total derivative / chain rule to
+            # simplify things. The final jacobian of the kinetics must be
+            # computed as jac(y) + jac(o) * observables_matrix.
+            self.kinetics_jacobian_y = kinetics_sparse.jacobian(self.y)
+            self.kinetics_jacobian_o = kinetics_sparse.jacobian(self.o)
+        self.observables_matrix = obs_matrix.tocsr()
+        self.stoichiometry_matrix = model.stoichiometry_matrix
+        self.model_name = model.name
+        self._expressions_constant = [
+            e.expand_expr().xreplace(param_subs) for e in expr_constant
+        ]
+        self._work_path = None
+        self._expressions_constant_fn = None
+        self._rhs_fn = None
+        self._jacobian_fn = None
+
+    def __del__(self):
+        if self._work_path is not None and self.cleanup:
+            shutil.rmtree(self._work_path, ignore_errors=True)
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        # Clear cached function objects (will be regenerated on demand).
+        state["_expressions_constant_fn"] = None
+        state["_rhs_fn"] = None
+        state["_jacobian_fn"] = None
+        return state
+
+    @property
+    def work_path(self):
+        """Location for saving any temporary working files."""
+        if self._work_path is None:
+            # Only initialize this on demand since it will create a directory
+            # which requires cleanup on destruction.
+            self._work_path = tempfile.mkdtemp(suffix="_pysb_compile")
```

Comment:
Might be good to log this path at `debug` level

Reply:
I might sprinkle a few more logging calls around, now that you mention it.